node/lib/util/merge_util.js

```diff
@@ -129,294 +141,518 @@ exports.fastForwardMerge = co.wrap(function *(repo, commit, message) {
 });
 
 /**
- * Merge the specified `subs` in the specified `repo` having the specified
- * `index`.  Stage submodule commits in `metaRepo`.  Return an object
- * describing any commits that were generated and conflicted commits.  Use the
- * specified `opener` to acces submodule repos.  Use the specified `message` to
- * write commit messages.
- * @param {NodeGit.Repository} metaRepo
- * @param {Open.Opener}        opener
- * @param {NodeGit.Index}      metaIndex
- * @param {Object}             subs        map from name to SubmoduleChange
+ * Get the left side commit and sha for submodule `subRepo` based on workding
+ * directory type.
+ * 
+ * @async
+ * @param {NodeGit.Repository}  subRepo
+ * @param {SubmoduleChange}     subChange
+ * @param {WDTYPE}              wdType
  * @return {Object}
- * @return {Object} return.commits    map from name to map from new to old ids
- * @return {Object} return.conflicts  map from name to commit causing conflict
+ * @return {String}             return.outSha
+ * @return {NodeGit.Commit}     return.outCommit
  */
-const mergeSubmodules = co.wrap(function *(repo,
-                                           opener,
-                                           index,
-                                           subs,
-                                           message) {
-    assert.instanceOf(repo, NodeGit.Repository);
-    assert.instanceOf(opener, Open.Opener);
-    assert.instanceOf(index, NodeGit.Index);
-    assert.isObject(subs);
-    assert.isString(message);
-
-    const result = {
-        conflicts: {},
-        commits: {},
+const leftCommitForSubmodule = co.wrap(function *(subRepo, subChange, wdType) {
+    if (WDTYPE.FORCE_BARE === wdType) {
+        return {
+            ourSha: subChange.ourSha,
+            ourCommit: yield subRepo.getCommit(subChange.ourSha),
+        };
+    } 
+    const headCommit = yield subRepo.getHeadCommit();
+    return {
+        ourSha: headCommit.id().tostrS(),
+        ourCommit: headCommit,
     };
-    const sig = yield ConfigUtil.defaultSignature(repo);
-    const fetcher = yield opener.fetcher();
-    const mergeSubmodule = co.wrap(function *(name) {
-        const subRepo = yield opener.getSubrepo(name, false);
-        const change = subs[name];
+});
 
-        const fromSha = change.newSha;
-        yield fetcher.fetchSha(subRepo, name, fromSha);
-        const subHead = yield subRepo.getHeadCommit();
-        const headSha = subHead.id().tostrS();
-        const fromCommit = yield subRepo.getCommit(fromSha);
+/**
+ * Write tree representation of the index to the disk, create a commit
+ * from the tree and update reference if needed.
```

Comment:
So, this is sort of odd.  Why not have the merge command update the ref (and the merge-bare command not)?  That is, try to avoid having a method that does two different things.

Reply:
Now it is the case that interactive merge update the ref ("HEAD") and the bare merge not. But interactive merge does more than that: 
1. it operates directly on repo's current index, so users can see all the file changes if merge is paused for user actions.
2. it pulls necessary changes when needed
3. it force reset subrepo's head when needed

Reply:
The first is just a question of what index is used, right?

The second of these should be true for both kinds of merge, right?  Unless by "pull", you really mean "pull" as in fetch+merge.  

Even if 1 and 2 do really both need to be conflated into this method, it does not follow that 3 does.  Our goal is clean code, which means separation of concerns as much as possible.

Reply:
Ok, let's move the logic to `MergeContext` then.